// @vkid/sdk/dist-sdk/cjs/core/config/config.js
'use strict';

var constants = require('../../constants.js');
var types = require('./types.js');

function _class_call_check(instance, Constructor) {
    if (!(instance instanceof Constructor)) {
        throw new TypeError("Cannot call a class as a function");
    }
}
function _defineProperties(target, props) {
    for(var i = 0; i < props.length; i++){
        var descriptor = props[i];
        descriptor.enumerable = descriptor.enumerable || false;
        descriptor.configurable = true;
        if ("value" in descriptor) descriptor.writable = true;
        Object.defineProperty(target, descriptor.key, descriptor);
    }
}
function _create_class(Constructor, protoProps, staticProps) {
    if (protoProps) _defineProperties(Constructor.prototype, protoProps);
    if (staticProps) _defineProperties(Constructor, staticProps);
    return Constructor;
}
function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
var Config = /*#__PURE__*/ function() {
    function Config() {
        _class_call_check(this, Config);
        _define_property(this, "store", {
            app: 0,
            redirectUrl: "",
            mode: types.ConfigAuthMode.Redirect,
            __loginDomain: constants.LOGIN_DOMAIN,
            __oauthDomain: constants.OAUTH_DOMAIN,
            __vkidDomain: constants.VKID_DOMAIN
        });
    }
    _create_class(Config, [
        {
            key: "set",
            value: function set(config) {
                this.store = _object_spread({}, this.store, config);
                return this;
            }
        },
        {
            key: "get",
            value: function get() {
                return this.store;
            }
        }
    ]);
    return Config;
}();

exports.Config = Config;
