// @vkid/sdk/dist-sdk/esm/widgets/floatingOneTap/floatingOneTap.js
import { validator } from '../../core/validator/validator.js';
import { isRequired } from '../../core/validator/rules.js';
import { Widget } from '../../core/widget/widget.js';
import '../../core/widget/events.js';
import { WidgetState } from '../../core/widget/types.js';
import { Languages, Scheme } from '../../types.js';
import { AgreementsDialog } from '../agreementsDialog/agreementsDialog.js';
import { AgreementsDialogInternalEvents } from '../agreementsDialog/events.js';
import { OAuthList } from '../oauthList/oauthList.js';
import { OAuthName } from '../oauthList/types.js';
import { FloatingOneTapInternalEvents } from './events.js';
import { getFloatingOneTapTemplate } from './template.js';
import { FloatingOneTapContentId } from './types.js';

function _ts_decorate(decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for(var i = decorators.length - 1; i >= 0; i--)if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
}
const defaultIndent = {
    top: 12,
    right: 12,
    bottom: 12
};
class FloatingOneTap extends Widget {
    vkidAppName = 'floating_one_tap_auth';
    onBridgeMessageHandler(event) {
        switch(event.handler){
            case FloatingOneTapInternalEvents.LOGIN_SUCCESS:
                {
                    this.redirectWithPayload(event.params);
                    break;
                }
            case FloatingOneTapInternalEvents.SHOW_FULL_AUTH:
                {
                    const params = {};
                    if (event.params.screen) {
                        params.screen = event.params.screen;
                    }
                    if (event.params.sdk_oauth) {
                        params.action = {
                            name: 'sdk_oauth',
                            params: {
                                oauth: event.params.sdk_oauth
                            }
                        };
                    }
                    this.openFullAuth(params);
                    break;
                }
            case FloatingOneTapInternalEvents.NOT_AUTHORIZED:
                {
                    this.setState(WidgetState.NOT_LOADED);
                    setTimeout(()=>{
                        // Ожидает выполнение анимации и меняет ui
                        this.setState(WidgetState.LOADED);
                    }, 500);
                    clearTimeout(this.timeoutTimer);
                    break;
                }
            case FloatingOneTapInternalEvents.SHOW_AGREEMENTS_DIALOG:
                {
                    this.createAgreementsDialogWidget();
                    break;
                }
            default:
                {
                    super.onBridgeMessageHandler(event);
                    break;
                }
        }
    }
    createAgreementsDialogWidget() {
        const params = {
            container: document.body,
            lang: this.lang,
            scheme: this.scheme
        };
        const agreementsDialog = new AgreementsDialog();
        const acceptHandler = (event)=>{
            this.bridge.sendMessage({
                handler: FloatingOneTapInternalEvents.START_AUTHORIZE,
                params: event.params
            });
            agreementsDialog.off(AgreementsDialogInternalEvents.ACCEPT, acceptHandler);
            agreementsDialog.close();
        };
        agreementsDialog.on(AgreementsDialogInternalEvents.ACCEPT, acceptHandler);
        agreementsDialog.render(params);
    }
    openFullAuth(value) {
        const params = {
            ...value,
            lang: this.lang,
            scheme: this.scheme
        };
        FloatingOneTap.__auth.login(params);
    }
    renderOAuthList(params) {
        if (!params.oauthList.length) {
            return;
        }
        const oauthList = new OAuthList();
        oauthList.render(params);
    }
    render(params) {
        this.lang = params?.lang || Languages.RUS;
        this.scheme = params?.scheme || Scheme.LIGHT;
        const providers = (params.oauthList || []).filter((provider)=>provider !== OAuthName.VK);
        const queryParams = {
            scheme: this.scheme,
            lang_id: this.lang,
            show_alternative_login: params?.showAlternativeLogin ? 1 : 0,
            content_id: params?.contentId || FloatingOneTapContentId.SIGN_IN_TO_SERVICE,
            providers: providers.join(',')
        };
        this.templateRenderer = getFloatingOneTapTemplate({
            openFullAuth: this.openFullAuth.bind(this),
            close: this.close.bind(this),
            scheme: this.scheme,
            lang: this.lang,
            indent: Object.assign(defaultIndent, params.indent || {}),
            contentId: queryParams.content_id,
            appName: params.appName,
            renderOAuthList: this.renderOAuthList.bind(this),
            providers
        });
        return super.render({
            container: document.body,
            ...queryParams
        });
    }
}
_ts_decorate([
    validator({
        appName: [
            isRequired
        ]
    })
], FloatingOneTap.prototype, "render", null);

export { FloatingOneTap };
