// @vkid/sdk/dist-sdk/cjs/utils/url.js
'use strict';

require('./../lib/@vkontakte/vkjs/lib/es6/detections.js');
var querystring = require('./../lib/@vkontakte/vkjs/lib/es6/querystring.js');
var constants = require('../constants.js');
var uuid = require('./uuid.js');

function _define_property(obj, key, value) {
    if (key in obj) {
        Object.defineProperty(obj, key, {
            value: value,
            enumerable: true,
            configurable: true,
            writable: true
        });
    } else {
        obj[key] = value;
    }
    return obj;
}
function _object_spread(target) {
    for(var i = 1; i < arguments.length; i++){
        var source = arguments[i] != null ? arguments[i] : {};
        var ownKeys = Object.keys(source);
        if (typeof Object.getOwnPropertySymbols === "function") {
            ownKeys = ownKeys.concat(Object.getOwnPropertySymbols(source).filter(function(sym) {
                return Object.getOwnPropertyDescriptor(source, sym).enumerable;
            }));
        }
        ownKeys.forEach(function(key) {
            _define_property(target, key, source[key]);
        });
    }
    return target;
}
function ownKeys(object, enumerableOnly) {
    var keys = Object.keys(object);
    if (Object.getOwnPropertySymbols) {
        var symbols = Object.getOwnPropertySymbols(object);
        if (enumerableOnly) {
            symbols = symbols.filter(function(sym) {
                return Object.getOwnPropertyDescriptor(object, sym).enumerable;
            });
        }
        keys.push.apply(keys, symbols);
    }
    return keys;
}
function _object_spread_props(target, source) {
    source = source != null ? source : {};
    if (Object.getOwnPropertyDescriptors) {
        Object.defineProperties(target, Object.getOwnPropertyDescriptors(source));
    } else {
        ownKeys(Object(source)).forEach(function(key) {
            Object.defineProperty(target, key, Object.getOwnPropertyDescriptor(source, key));
        });
    }
    return target;
}
var getVKIDUrl = function(module, params, config) {
    var queryParams = _object_spread_props(_object_spread({}, params), {
        uuid: params.uuid || uuid.uuid(),
        v: constants.VERSION,
        sdk_type: "vkid",
        app_id: config.app,
        redirect_uri: config.redirectUrl,
        redirect_state: config.state,
        debug: config.__debug ? 1 : null,
        localhost: config.__localhost ? 1 : null
    });
    var queryParamsString = querystring.querystring.stringify(queryParams, {
        skipNull: true
    });
    return "https://".concat(config.__vkidDomain, "/").concat(module, "?").concat(queryParamsString);
};
// TODO: добавить типы
var getRedirectWithPayloadUrl = function(payload, config) {
    var url = "".concat(config.get().redirectUrl, "?payload=").concat(encodeURIComponent(JSON.stringify(payload)));
    if (config.get().state) {
        url += "&state=".concat(config.get().state);
    }
    return url;
};

exports.getRedirectWithPayloadUrl = getRedirectWithPayloadUrl;
exports.getVKIDUrl = getVKIDUrl;
